File: S03/week05/sejun/ClimbingStairs.java
// 0ms Beats 100.00% of users with Java
// Memory 40.10MB Beats 73.81% of users with Java
class Solution {
    int[] dp = new int[46];
    public int climbStairs(int n) {
        dp[0] = 1;
        for (int i = 1; i <= n; i++) {
            int oneStep = dp[i-1];
            int twoStep = 0;
            if (i -2 >= 0) {
                twoStep = dp[i-2];
            }

            dp[i] = oneStep + twoStep;
        }

        return dp[n];
    }
}

File: S03/week05/z0/ClimbingStairs.kt
class Solution {
    // 123ms(56.97%)
    // 32.91MB(65.49%)
    fun climbStairs(n: Int): Int {
        val array = IntArray(n+1)
        array[0] = 1
        array[1] = 1

        // f(n-2) + f(n-1) = f(n)
        for(i in 2..n) {
            array[i] = array[i-1] + array[i-2]
        }
        return array[n]
    }
}
